backend/cfg/cfg_cse.ml

```diff
@@ -0,0 +1,250 @@
+                   very cheap to compute, e.g. an integer constant, don't
+                   bother. *)
+                let n3 = set_known_regs n1 i.res vres in
+                (* This is n1 above and not n2 because the move does not destroy
+                   any regs *)
+                insert_move state res i.res cell;
+                DLL.delete_curr cell;
+                n3
+              | _ ->
+                (* We already computed the operation but lost its results.
+                   Associate the result registers to the result valnums of the
+                   previous operation. *)
+                let n3 = set_known_regs n2 i.res vres in
+                n3)
+            | None ->
+              (* This operation produces a result we haven't seen earlier. *)
+              let n3 = set_fresh_regs n2 i.res (op, varg) op_class in
+              n3)
+          | Op_store false | Op_other ->
+            (* An initializing store or an "other" operation do not invalidate
+               any equations, but we do not know anything about the results. *)
+            let n1 = set_unknown_regs n (Proc.destroyed_at_basic i.desc) in
+            let n2 = set_unknown_regs n1 i.res in
+            n2
+          | Op_store true ->
+            (* A non-initializing store can invalidate anything we know about
+               prior mutable loads. *)
+            let n1 = set_unknown_regs n (Proc.destroyed_at_basic i.desc) in
+            let n2 = set_unknown_regs n1 i.res in
+            let n3 = self#kill_loads n2 in
+            n3)
+    [@@warning "-4"]
+
+    method private cse_body
+        : State.t -> numbering -> Cfg.basic Cfg.instruction DLL.t -> numbering =
+      fun state numbering body ->
+        let numbering = ref numbering in
+        DLL.iter_cell body
+          ~f:(fun (cell : Cfg.basic Cfg.instruction DLL.cell) ->
+            numbering := self#cse_instruction state !numbering cell);
+        !numbering
+
+    method private cse_terminator
+        : numbering -> Cfg.terminator Cfg.instruction -> numbering =
+      fun numbering terminator ->
+        match terminator.desc with
+        | Never -> assert false
+        | Always _ | Parity_test _ | Truth_test _ | Float_test _ | Int_test _
+        | Switch _ ->
+          set_unknown_regs numbering
+            (Proc.destroyed_at_terminator terminator.desc)
+        | Return | Raise _ | Tailcall_self _ | Tailcall_func _
+        | Call_no_return _ | Call _ | Prim _ ->
+          empty_numbering
+        | Specific_can_raise _ ->
+          (* CR-soon xclerc for xclerc: is it too conservative? *)
+          empty_numbering
+
+    method private cse_blocks : State.t -> Cfg.t -> unit =
+      fun state cfg ->
+        let visited = ref Label.Set.empty in
+        let to_visit : (numbering * Cfg.basic_block) Queue.t =
+          Queue.create ()
+        in
+        (* blocks with zero or several predecessors and exception handlers start
+           with an empty numbering *)
```

Comment:
To me, it is actually implied by the comment
because a join point has several predecessors.
I have amended the comment to be more explicit;
also copied the missing comments.

Reply:
Thank you! The reason I thought it is not implied when reading the PR is that there can be some sort of a fixpoint in propagating numbering (there is no fixpoint, just one pass, neither here not in the upstream version).